### PR TITLE
Convert RDD[Vector]/RDD[LabeledPoint] to HomogenTable

### PR DESCRIPTION
**Pull request by minmingzhu**
## What changes were proposed in this pull request?

1. convert RDD[Vector]/RDD[LabeledPoint] to HomogenTable
2. add unit test

## Does this PR also require the following changes?

- CI
- Documentation
- Example

### CODE REVIEW

File: mllib-dal/src/main/native/oneapi/dal/HomogenTableImpl.cpp [hunk 1]
```diff
@@ -424,3 +419,56 @@ Java_com_intel_oneapi_dal_table_HomogenTableImpl_cGetDoubleData(
...
+/*
+ * Class:     com_intel_oneapi_dal_table_HomogenTableImpl
+ * Method:    cEmptyTableInit
+ * Signature: ()J
+ */
+JNIEXPORT jlong JNICALL Java_com_intel_oneapi_dal_table_HomogenTableImpl_cEmptyTableInit
+  (JNIEnv *env, jobject) {
+      printf(" init empty HomogenTable \n");
+      homogen_table *h_table = new homogen_table();
+      homogenPtr tablePtr = std::make_shared<homogen_table>(*h_table);
```
Comment (xwu99) [inline mllib-dal/src/main/native/oneapi/dal/HomogenTableImpl.cpp:432]: Line 431-432 can be merged into single one. 
std::make_shared is to new the object and make share_ptr at the same time

```suggestion
      homogenPtr tablePtr = std::make_shared<homogen_table>();
```

Comment (xwu99) [inline mllib-dal/src/main/native/oneapi/dal/HomogenTableImpl.cpp:432]: Line 431-432 can be merged into single one. 
std::make_shared is to new the object and make share_ptr at the same time

```suggestion
      homogenPtr tablePtr = std::make_shared<homogen_table>();
```

File: mllib-dal/src/main/native/oneapi/dal/HomogenTableImpl.cpp [hunk 2]
```diff
@@ -424,3 +419,56 @@ Java_com_intel_oneapi_dal_table_HomogenTableImpl_cGetDoubleData(
...
+              for (std::int64_t i = 0; i < mergeDatasize; i++) {
+                  row_doubles[i] = mergeData[i];
+              }
+              for (std::int64_t i = 0; i < homogenDatasize; i++) {
+                  row_doubles[mergeDatasize + i] = homogenData[i];
+              }
+
+              homogen_table *resultTable = new homogen_table(
+                      row_doubles, cRowCount, cColCount, detail::empty_delete<const double>());
+              homogenPtr resultTablePtr = std::make_shared<homogen_table>(*resultTable);
```
Comment (xwu99) [inline mllib-dal/src/main/native/oneapi/dal/HomogenTableImpl.cpp:468]: Same as above, should not use new.
```suggestion
              homogenPtr resultTablePtr = std::make_shared<homogen_table>(row_doubles, cRowCount, cColCount, detail::empty_delete<const double>());
```
```suggestion
              homogenPtr resultTablePtr = std::make_shared<homogen_table>(*resultTable);
```

Comment (xwu99) [inline mllib-dal/src/main/native/oneapi/dal/HomogenTableImpl.cpp:468]: Same as above, should not use new.
```suggestion
              homogenPtr resultTablePtr = std::make_shared<homogen_table>(row_doubles, cRowCount, cColCount, detail::empty_delete<const double>());
```
```suggestion
              homogenPtr resultTablePtr = std::make_shared<homogen_table>(*resultTable);
```

File: mllib-dal/src/main/native/oneapi/dal/HomogenTableImpl.cpp [hunk 3]
```diff
@@ -424,3 +419,56 @@ Java_com_intel_oneapi_dal_table_HomogenTableImpl_cGetDoubleData(
...
+              const double* mergeData = mergeTable.get_data<double>();
+              const int mergeDatasize = mergeTable.get_column_count() * mergeTable.get_row_count();
+
+
+              const double* homogenData = homogenTable.get_data<double>();
+              const int homogenDatasize = homogenTable.get_column_count() * homogenTable.get_row_count();
+
+              long cRowCount = mergeTable.get_row_count() + homogenTable.get_row_count();
+              long cColCount = mergeTable.get_column_count();
+              double* row_doubles = new double[mergeDatasize + homogenDatasize];
```
Comment (xwu99) [inline mllib-dal/src/main/native/oneapi/dal/HomogenTableImpl.cpp:458]: should not use new

Comment (xwu99) [inline mllib-dal/src/main/native/oneapi/dal/HomogenTableImpl.cpp:458]: should not use new

File: mllib-dal/src/main/java/com/intel/oneapi/dal/table/RowAccessor.java [hunk 4]
```diff
@@ -3,7 +3,6 @@
 public class RowAccessor {
     private long cObject;
     private Common.ComputeDevice cDevice;
-
     public RowAccessor(long cObject, Common.ComputeDevice device) {
         this.cObject = cObject;
         this.cDevice =device;
```
Comment (xwu99) [inline mllib-dal/src/main/java/com/intel/oneapi/dal/table/RowAccessor.java:8]:
```suggestion
        this.cDevice = device;
```

Comment (xwu99) [inline mllib-dal/src/main/java/com/intel/oneapi/dal/table/RowAccessor.java:8]:
```suggestion
        this.cDevice = device;
```
```suggestion
        this.cDevice =device;
```

File: mllib-dal/src/main/native/oneapi/dal/HomogenTableImpl.cpp [hunk 5]
```diff
@@ -235,8 +235,8 @@ JNIEXPORT jlong JNICALL Java_com_intel_oneapi_dal_table_HomogenTableImpl_lInit(
     jint cLayout, jint cComputeDevice) {
     printf("HomogenTable long init \n");
     jlong *fData = env->GetLongArrayElements(cData, NULL);
-    homogen_table *h_table ;
-    homogenPtr tablePtr ;
+    homogen_table *h_table;
+    homogenPtr tablePtr;
```
Comment (xwu99) [inline mllib-dal/src/main/native/oneapi/dal/HomogenTableImpl.cpp:239]: there are code duplication below. pls refactor.

Comment (xwu99) [inline mllib-dal/src/main/native/oneapi/dal/HomogenTableImpl.cpp:239]: there are code duplication below. pls refactor.

File: mllib-dal/src/main/scala/com/intel/oap/mllib/OneDAL.scala [hunk 6]
```diff
@@ -428,6 +425,78 @@ object OneDAL {
...
+    val coalescedTables = tables.rdd.coalesce(executorNum,
+      partitionCoalescer = Some(new ExecutorInProcessCoalescePartitioner()))
+
+    val mergedTables = coalescedTables.mapPartitions { iter =>
+      val mergedFeatures = new HomogenTable()
+      val mergedLabels = new HomogenTable()
+
+      iter.foreach { case (featureAddr, labelAddr) =>
+        mergedFeatures.addHomogenTable(featureAddr)
+        mergedLabels.addHomogenTable(labelAddr)
```
Comment (xwu99) [inline mllib-dal/src/main/scala/com/intel/oap/mllib/OneDAL.scala:490]: If the merging copied discrete buffer into continious buffer, the original buffer should be released to save memory.

Comment (xwu99) [inline mllib-dal/src/main/scala/com/intel/oap/mllib/OneDAL.scala:490]: If the merging copied discrete buffer into continious buffer, the original buffer should be released to save memory.

File: mllib-dal/src/test/scala/com/intel/oap/mllib/ConvertHomogenTableSuite.scala [hunk 7]
```diff
@@ -94,10 +94,10 @@ class ConvertHomogenTableSuite extends FunctionsSuite with Logging {
     val data = Array(5.308206,9.869278,1.018934,4.292158,6.081011,6.585723,2.411094,4.767308,-3.256320,-6.029562)
 
     val expectData = Array(5.308206,9.869278)
-
     val table = new HomogenTable(5, 2, data, getDevice)
     val vector = OneDAL.homogenTable1xNToVector(table, getDevice)
 
+
     assert(expectData=== vector.toArray)
```
Comment (xwu99) [inline mllib-dal/src/test/scala/com/intel/oap/mllib/ConvertHomogenTableSuite.scala:101]: should use assertArrayEquals?

Comment (xwu99) [inline mllib-dal/src/test/scala/com/intel/oap/mllib/ConvertHomogenTableSuite.scala:101]: should use assertArrayEquals?

**Review by xwu99 — CHANGES_REQUESTED** [review]

File: mllib-dal/src/test/scala/com/intel/oap/mllib/ConvertHomogenTableSuite.scala (hunk 7 above)
Comment (minmingzhu) [reply mllib-dal/src/test/scala/com/intel/oap/mllib/ConvertHomogenTableSuite.scala:101]: Scala has not assertArrayEquals function.

Comment (minmingzhu) [reply mllib-dal/src/test/scala/com/intel/oap/mllib/ConvertHomogenTableSuite.scala:101]: Scala has not assertArrayEquals function.

File: mllib-dal/src/main/native/oneapi/dal/SimpleMetadataImpl.cpp [hunk 8]
```diff
@@ -66,7 +66,7 @@ JNIEXPORT jint JNICALL
 Java_com_intel_oneapi_dal_table_SimpleMetadataImpl_cGetDataType(
     JNIEnv *env, jobject, jlong cTableAddr, jint cindex) {
     printf("SimpleMetadata getdatatype \n");
-    table_metadata mdata =
-        *((table_metadata *)cTableAddr);
+    table_metadata mdata = *((table_metadata *)cTableAddr);
```
Comment (xwu99) [inline mllib-dal/src/main/native/oneapi/dal/SimpleMetadataImpl.cpp:69]: could you use c++ style type casting (reinterpret_cast/static_cast etc.)? same for other places.

File: mllib-dal/src/test/scala/org/apache/spark/ml/oneDALSuite.scala [hunk 9]
```diff
@@ -37,4 +45,94 @@ class oneDALSuite extends FunctionsSuite with Logging {
...
+    }
+    data
+  }
+
+  private def getDevice: Common.ComputeDevice = {
+    val device = System.getProperty("computeDevice")
+    var computeDevice: Common.ComputeDevice = Common.ComputeDevice.HOST
+    if(device != null) {
+      device.toUpperCase match {
+        case "HOST" =>  computeDevice = Common.ComputeDevice.HOST
```
Comment (xwu99) [inline mllib-dal/src/test/scala/org/apache/spark/ml/oneDALSuite.scala:126]:
```suggestion
        case "HOST" => computeDevice = Common.ComputeDevice.HOST
```

File: mllib-dal/src/test/scala/org/apache/spark/ml/oneDALSuite.scala [hunk 10]
```diff
@@ -37,4 +45,94 @@ class oneDALSuite extends FunctionsSuite with Logging {
...
+        case "CPU"  => computeDevice = Common.ComputeDevice.CPU
```
Comment (xwu99) [inline mllib-dal/src/test/scala/org/apache/spark/ml/oneDALSuite.scala:127]:
```suggestion
        case "CPU" => computeDevice = Common.ComputeDevice.CPU
```

File: mllib-dal/src/test/scala/org/apache/spark/ml/oneDALSuite.scala [hunk 11]
```diff
@@ -37,4 +45,94 @@ class oneDALSuite extends FunctionsSuite with Logging {
...
+        case "GPU"  => computeDevice = Common.ComputeDevice.GPU
```
Comment (xwu99) [inline mllib-dal/src/test/scala/org/apache/spark/ml/oneDALSuite.scala:128]:
```suggestion
        case "GPU" => computeDevice = Common.ComputeDevice.GPU
```

File: mllib-dal/src/test/scala/org/apache/spark/ml/oneDALSuite.scala [hunk 12]
```diff
@@ -37,4 +45,94 @@ class oneDALSuite extends FunctionsSuite with Logging {
...
+        case _  => "Invalid Device"
```
Comment (xwu99) [inline mllib-dal/src/test/scala/org/apache/spark/ml/oneDALSuite.scala:129]:
```suggestion
        case _ => "Invalid Device"
```

File: mllib-dal/src/test/scala/com/intel/oap/mllib/ConvertHomogenTableSuite.scala (hunk 7 above)
Comment (xwu99) [reply mllib-dal/src/test/scala/com/intel/oap/mllib/ConvertHomogenTableSuite.scala]: this is a JUnit function, you should be able to use it in Scala

**Review by xwu99 — CHANGES_REQUESTED** [review]

**Comment by xwu99** [comment]
Thanks, could you rebase #7 and I will review it. thanks!

**Pull request merged**